### ISSUE TITLE
Testcafe 0.23.2 hanging once tests have completed

### ISSUE THREAD

**Issue by grittershub** [DevExpress/testcafe#3114]
### What is your Test Scenario?

 I'm running testcafe from a codebuild docker image based on ubuntu. Once all tests are completed, it does not return, it just hangs indefinitely. Commandline, test file and logs are below. Testfile will run fine from a macbook locally.

Testcafe is version 0.23.2
### What is the Current behavior?

 Testcafe never returns once the tests have completed, successfully or otherwise
### What is the Expected behavior?

 Testcafe to return

### What is your web application and your TestCafe test code?
Logs:

```
Tue, 13 Nov 2018 09:51:22 GMT testcafe:test-run:HeadlessChrome
70.0.3538 / Linux 0.0.0:driver-message  
    { 
      "cmd": "ready", 
      "status": { 
        "id": "2829.9000000006345", 
        "isCommandResult": true, 
        "executionError": null, 
        "pageError": null, 
        "resent": false, 
        "result": [ 
          0 
        ], 
        "consoleMessages": { 
          "log": [], 
          "info": [], 
          "warn": [], 
          "error": [] 
        } 
      }, 
      "disableResending": true, 
      "allowRejecting": true, 
      "sessionId": "J07WRZJDH" 
    }    Tue, 13 Nov 2018 09:51:22 GMT testcafe:test-run:HeadlessChrome 70.0.3538 / Linux 0.0.0:command  
    { 
      "type": "test-done" 
    }    Tue, 13 Nov 2018 09:51:22 GMT testcafe:test-run:HeadlessChrome 70.0.3538 / Linux
0.0.0:driver-message  
    { 
      "cmd": "ready", 
      "status": { 
        "id": "2851.400000000467", 
        "isCommandResult": true, 
        "executionError": null, 
        "pageError": null, 
        "resent": false, 
        "result": null, 
        "consoleMessages": { 
          "log": [], 
          "info": [], 
          "warn": [], 
          "error": [] 
        } 
      }, 
      "disableResending": true, 
      "allowRejecting": true, 
      "sessionId": "J07WRZJDH" 
    }     ✓ Login nogroup user 
     5 passed (31s)

```

### Steps to Reproduce:

Run with the commandline above using a simple test file. Test file has a single fixture and 5 test cases. The tests themselves actually complete successfully and testcafe returns 'passed' at the end of the test.

```

import { Selector } from 'testcafe'; import { Role } from 'testcafe'; import { ClientFunction } from 'testcafe';

fixture `Login`
    .page `url goes here`;

const getPageUrl = ClientFunction(() => window.location.href);

test('Login user', async t => {     await t  ..test example code here... });

```
### Your Environment details:
 
* testcafe version:                   

```
root@3c9892b7d478:/tmp# testcafe -v
0.23.2<!-- run `testcafe -v` -->

```

* node.js version:  
                  
```
root@3c9892b7d478:/tmp# node -v
v10.1.0<!-- run `node -v` -->

```
* command-line arguments:     
`export DEBUG="testcafe:test-run:* gulp test-functional"; testcafe "chrome:headless --no-sandbox" test/client/login.js 
`
* browser name and version: 
`Google Chrome 70.0.3538.102
`
* platform and version:         
`Ubuntu 14.04.5 LTS 
`
* other:

**Comment by AndreyBelym** [comment]
Could you please provide additional resources to help me reproduce and examine the issue:
 - [ ] Your `Dockerfile` or a link to your image on Docker Hub
 - [ ] The full code of your test files
 - [ ] If you can't disclose your tests or site, please try to reproduce the issue on another public site (`example.com` `google.com`, etc.) and provide a test that uses this site.

**Comment by grittershub** [comment]
Confirmed that rolling testcafe back to v0.21.1 results in expected behaviour, tests complete and testcafe now terminates successfully with no changes to any other configuration

**Comment by dlangerenken** [comment]
I have a similar issue in a Jenkins/docker environment that runs testcafe in the CI gate. It also hang and downgrading to 0.21.1 helped as well.

**Comment by AndreyBelym** [comment]
I've built the image corresponding to the `Dockerfile` configuration you provided and tried to run it manually with the following command:
```
docker run -it --rm -v /d/gh-3114:/d/ andreybelym/test-images:gh-3114 'export DEBUG="testcafe:test-run:* gulp test-functional"; testcafe "chrome:headless --no-sandbox"
 /d/test.js'
```

I ran the image and it was terminated successfully. I've pushed it to Docker Hub, so could you please try to run it on your machine? You can get it by running the following command
```
docker pull andreybelym/test-images:gh-3114 
```

I see that you use AWS CodeBuild to run your builds and expect that the problem can be reproducible in the CodeBuild environment but I need additional time to set it up.

 I also checked commits we  introduced between the working (`v0.23.1`) and problematic (`v0.23.2`) TestCafe versions and found that we had only 2 commits that could cause the issue: [6f8db5ec](https://github.com/DevExpress/testcafe/commit/6f8db5ec9be84aaca23ccb340fe352b2a7c993e6) and [f105949a](https://github.com/DevExpress/testcafe/commit/f105949a8cf5de6ddc27fa51fb2e94eccc298824).
 
So I've prepared debug builds for both of these commits, and also a third build that is just a rebuilt `v0.23.1` version and will serve as a control build for checking that the issue isn't caused by some automatically updated TestCafe dependency. Below is the list of links to these builds:
 - https://github.com/AndreyBelym/testcafe/releases/download/gh-3114-1/testcafe-0.23.3-alpha.gh3114.1.tgz
- https://github.com/AndreyBelym/testcafe/releases/download/gh-3114-2/testcafe-0.23.3-alpha.gh3114.2.tgz
- https://github.com/AndreyBelym/testcafe/releases/download/gh-3114-3/testcafe-0.23.3-alpha.gh3114.3.tgz

While I'm setting the environment to reproduce the problem, could you try to install these builds to help me determine the exact commit that caused the issue?
  
You can install any of these builds by specifying the following line in the `dependencies` section of your `package.json`:
```
"testcafe": "LINK"
```
or by executing the following command in your terminal:
```
npm i -g LINK
```

Replace `LINK` with the actual link to a build.

Also, please modify your build steps to include the output of the `testcafe -v` command. It will help you check that a correct TestCafe version has been installed.

**Comment by grittershub** [comment]
Another point to note, 0.23.3-alpha.gh3114.1 (in fact all 3 versions) execute successfully when run locally within the docker image, with the hanging only occurring when run from the CodeBuild execution environment..

**Comment by AlexKamaev** [comment]
@grittershub 
We prepared a build with a fix for you. Could you please check if it solves the issue? You can download it from the following link https://github.com/AndreyBelym/testcafe/releases/download/gh-3114-4/testcafe-0.23.3-alpha.gh3114.4.tgz

**Comment by LavrovArtem** [comment]
@grittershub
The fix was published in `testcafe@0.23.3-alpha.1`.

**Comment by willscripted** [comment]
Looks like I have the same problem with `0.23.3` when passing in an `outStream` to a reporter. Appears to be an order-of-operations issues [here](https://github.com/DevExpress/testcafe/blob/v0.23.3-alpha.1/src/reporter/index.js#L151-L156)

```
this.outStream.end();

await new Promise(resolve => {
     this.outStream.once('finish', resolve);
     this.outStream.once('error', resolve);
});
```

If `this.outStream.end()` fires the `finish` event synchronously, the promise will never resolve. The `once` handler is not called for passed events.

The fix should be pretty straightforward; just create the promise before `this.outStream.end()`

```
const p = new Promise(resolve => {
     this.outStream.once('finish', resolve);
     this.outStream.once('error', resolve);
});

this.outStream.end();

await p;
```

**Comment by AndreyBelym** [comment]
@willscripted, the fix for the described problem is in PR #3234.
